components_library/memory/abstract_memory_system.py

```diff
@@ -41,17 +41,33 @@ def __init__(self) -> None:
 
     @abstractmethod
     def incorporate_memory(self, board: AbstractBoard) -> None:
+        """This function completes all of the necessary steps to add this
+        memory system to the board."""
         raise NotImplementedError
 
     @abstractmethod
     def get_mem_ports(self) -> Sequence[Tuple[AddrRange, Port]]:
+        """Get the ports to connect this memory system to the cache"""
         raise NotImplementedError
 
     @abstractmethod
     def get_memory_controllers(self) -> List[MemCtrl]:
+        """Get all of the memory controllers in this memory system"""
         raise NotImplementedError
 
     @abstractmethod
-    def get_memory_ranges(self) -> List[AddrRange]:
-        """All of the memory ranges that this memory system responds to."""
+    def get_size(self) -> int:
+        """Returns the total size of the memory system"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def set_memory_range(self, ranges: List[AddrRange]) -> None:
```

Comment:
I see why you've done it this way, but I deeply dislike having to set the memory ranges in this way. At the core of my design is that all the board-component interaction should happen with the "incorporate_x" functions. Do you think there's a way you could migrate that there? Even then I'm not super happy with that as it's something someone might forget about :/

A more complex design to enforce this could be:

```
def incorporate_memory(...):
   _incorporate_memory(...)
   _set_memory_range(...)

@abstractmemory
def _incorporate_memory(...):
   pass

@abstractmethod
def _set_memory_range(...)
   pass
```

in the AbstractMemorySystem.

Just a thought, willing to discuss this.

Reply:
Done!